README.md
All is down!
============

Minimalistic monitoring system with web output. Can check remote hosts availability via HTTP, HTTPS, ICMP echo request or TCP port connection. May be used as status page for some web- or related services. Also can send notifications in case of some service become unavailable.


System requirements
-------------------

**Linux** or other UNIX system (not tested, use at your own risc)

**Ruby 1.8** or newer with installer **Rubygems**


Installation
------------

1. Clone the repo with actual version using `git clone https://github.com/Mendor/allisdown.git`
2. Run `./geminstall.sh` to download and install latest versions of required Ruby gems.


Running
-------

The script `runscript.sh` in the root directory of _allisdown_ used for execution management. Just run it with the parameter `start`, `stop`, `restart` or `kill` according to necessary action.

The monitoring data is accessible by URL [http://127.0.0.1:6006/](http://127.0.0.1:6006/). Also there is XML data available at [http://127.0.0.1:6006/xml](http://127.0.0.1:6006/xml).


Configuration
-------------

All configuration files are stored under `conf/` directory. Further this directory name is skipped in configuration files decrtiption. All of them are using [YAML syntax](http://www.yaml.org/).

### Thin web server
By default Thin listens to port 6006 on interface 127.0.0.1. You may change these and some other settings in the file `thin.conf.yml`. Be careful during editing this file, incorrect settings may lead to web server's misfunctions.

### Global monitoring configuration
Configuration data for monitoring system is being read from the file `monitor.conf.yml`. Here are the parameters could be set there:
* _statusfile_ — lock-file used by monitoring (default value: `tmp/monitor.active`)
* _mainlog_ — main log file of monitoring (default value: `log/monitor.log`)
* _timeout_ — timeout between attempts to re-run checks (default value: `5`); be aware that every host should have its own re-check timeout, it's mentioned onward
* _hosts_ — monitored hosts configuration file (default value: `conf/hosts.conf.yml`)
* _frontend_ — output file used to generate frontend content (default value: `tmp/data.yml`)
* _notifications_ – block for settings for notifications via external services e.g. email or HTTP API; the suboption `default` means global notifications activity, don't forget to set notifications method with appropriate options for it

In case you need to re-locate main configuration file, you should also update `allisdown.rb` and `monitor.rb` with the new configuration location.

### Hosts configuration
By defaults, hosts configuration is being loaded from file `hosts.conf.yml`. Every YAML key under the root is configuration for one check of one node.

Common configuration example:
```
Check_ID:                        # ID must be unique
  type: http|https|ping|port     # type of check; required
  host: hostname_or_ip           # valid hostname or IP; required
  timeout: 60                    # timeout between this host's checks; fixed number; required
  description: any text          # used only for explanations at frontend; optional
  notifications: on              # override global notification activity for the specific host; yes/no; optional
```

According to chosen check type, you may need to set additional parameters:

* **http, https:**

  ``port: 80     # port for HTTP/HTTPS connection; optional; by defaults used 80 or 443 respectively``

* **port**

  ``port: 22     # TCP port to check its availability; required``
  
* **ping**

  ``iters: 5     # number of ICMP packets should be sent to remote host; required``
  
### Frontend
You may redesign a bit output page by editing `views/index.erb` template. It is easy.

Using with Docker
-----------------

On your Docker host (using root account):
```shell
git clone http://github.com/Mendor/allisdown.git
cd allisdown
cp -r conf /etc/allisdown
vim /etc/allisdown/hosts.conf.yml
docker build -t allisdown .
docker run -d -i -p 6006:6006 -v /etc/allisdown:/opt/allisdown/conf --restart=always -t allisdown
```

And check did it run correctly using ``curl http://localhost:6006/``.  

License
-------

All this project's source codes licensed under [WTFPL](http://sam.zoy.org/wtfpl/), but I would be grateful for your contribution whatever that may mean :).


CHANGELOG.md
**2015-03-04**
- Added notifications support
- Dockerized

**2013-01-15**
- Added XML output support

**2012-12-26**
- First public version, doing what can do.


src/allisdown/monitor.rb
#!/usr/bin/env ruby -wKU
# encoding: utf-8

# # # Main class for monitoring module

require 'yaml'
require 'date'
require 'fileutils'

$LOAD_PATH.unshift("#{File.dirname(__FILE__)}")
require 'check.rb'     # source of the module Check
require 'validate.rb'  # source of the module Validate
require 'notify.rb'    # notifications class

############   HERE BE DRAGONS   ############

class Monitor
  # default configuration values
  @@CONFIG     = 'conf/monitor.conf.yml'  # main monitoring configuration
  @@FRONTEND   = 'tmp/data.yml'           # file with check results used by frontend
  @@MAINLOG    = 'log/monitor.log'        # main log with all events
  @@HOSTS      = 'conf/hosts.conf.yml'    # file with hosts configuration
  @@TIMEOUT    = 5                        # timeout between check attempts iterations
  @@STATUSFILE = 'tmp/monitor.active'     # lock-file showing monitoring activity
  @@NOTIFY     = {'default' => 'off'}     # default settings for notifications
  # hosts data
  @@HOSTDATA   = ''
  @@TOTAL      = 0

  # private methods section
  private

  ## Writing log events
  def self.log(content)
    time = DateTime.now.strftime("%F %T %z")
    File.open(@@MAINLOG, 'a') do |logfile|
      logfile.puts "#{time} - #{content}"
    end
  end

  ## Generate output for frontend
  def self.gendata(type, content)
    if type == 'error'
      # if we got an error, we need also to log it
      self.log("ERROR: #{content}")
      File.open(@@FRONTEND, 'w') do |frontend|
        frontend.puts ['error' => content].to_yaml
      end
    else
      File.open(@@FRONTEND, 'w') do |frontend|
        frontend.puts content.to_yaml
      end
    end
  end

  ## Load configuration data
  def self.config_load?
    if !File.exist?(@@CONFIG)
      self.gendata('error', "Unable load configuration file #{@@CONFIG}")
      return false
    end
    config = YAML.load_file(@@CONFIG)
    # ====== data read — start ======
    @@STATUSFILE = config['statusfile']
    @@FRONTEND   = config['frontend']
    @@HOSTS      = config['hosts']
    @@MAINLOG    = config['mainlog']
    @@TIMEOUT    = config['timeout']
    @@NOTIFY     = config['notifications']
    # ====== data read — end ======
    self.log("Loaded configuration file #{@@CONFIG}")
    true
  end

  ## Validate hosts' configuration
  def self.hosts_validate(hostdata)
    hostdata.each do |key, host|
      # by defaults, current host is valid
      host_valid = true
      bad_params = ""
      # 'host' and 'timeout' keys are required for all check types
      if !Validate.host?(host['host'])
        host_valid = false
        bad_params += " host"
      end
      if !Validate.timeout?(host['timeout'])
        host_valid = false
        bad_params += " timeout"
      end
      case host['type']
      when 'http'
        # for HTTP we optionaly require port number
        if (!Validate.port?(host['port'], false))
          host_valid = false
          bad_params += " port"
        end
      when 'https'
        # for HTTPS we optionaly require port number
        if (!Validate.port?(host['port'], false))
          host_valid = false
          bad_params += " port"
        end
      when 'ping'
        # for ping we strongly require number of iterations
        if (!Validate.port?(host['iters'], true))
          host_valid = false
          bad_params += " iters"
        end
      when 'port'
        # for port we strongly require port number
        if (!Validate.port?(host['port'], true))
          host_valid = false
          bad_params += " port"
        end
      else
        host_valid = false
      end
      # I don't think we need to validate description
      if !host_valid
        self.gendata('error', "Bad configuration for host: #{key} (options:#{bad_params})")
        hostdata.delete(key)
      end
    end
  end

  ## Load hosts data
  def self.hosts_load?
    if !File.exist?(@@HOSTS)
      self.gendata('error', "Unable load hosts' configuration from #{@@HOSTS}")
      return false
    end
    @@HOSTDATA = YAML.load_file(@@HOSTS)
    # validate received configuration
    self.hosts_validate(@@HOSTDATA)
    # total hosts count
    @@TOTAL = @@HOSTDATA.size
    self.log("Loaded hosts' data from file #{@@HOSTS}, totally #{@@TOTAL} entries")
    true
  end

  ## Run check for chosen host
  def self.check(hostdata)
    host = hostdata['host']  # parameter 'host' is common for all check types
    case hostdata['type']
    when 'http'
      port = hostdata['port'] || 80     # 80 — default HTTP port
      Check.http(host, port)
    when 'https'
      port = hostdata['port'] || 443    # 443 — default HTTPS port
      Check.https(host, port)
    when 'ping'
      timeout = hostdata['iters'] || 5  # by default, we'll send 5 ICMP packets
      Check.ping(host, timeout)
    when 'port'
      port = hostdata['port']           # no any default value, how-ow
      Check.port(host, port)
    end
    # By default, all the checks returns "OK" if no errors, and some another information if check
    # failed.
    #
    # I don't want to pass Check class methods to meta-code section, because, possible, we need to
    # implement some check with more than 2 parameters. Or completely use another module for some
    # of them.
  end

  ## Main watching loop
  def self.watch!
    loop do
      # we don't need monitor. if lock-file is absent
      if !File.exist?(@@STATUSFILE)
        break
      end
      @@HOSTDATA.each do |host|
        # previous check timestamp
        lastcheck = host[1]['lastcheck'] || 0
        # get current timestamp
        now = Time.now.to_i
        # if the previous check has been before current time minus checking timeout for the host
        if now - lastcheck > host[1]['timeout']
          checked = self.check(host[1])  # check result
          # log event if host status has been changed
          if host[1]['status'] != checked
            self.log("UPDATE: #{host[0]}: #{checked}")
            host_notifications = host[1]['notifications'] || 'off'
            if @@NOTIFY['default'] or host_notifications == 'on'
              unless checked == 'OK' and @@NOTIFY['send_ok']
                n = Notify.new(@@NOTIFY)
                n.send(@@NOTIFY['method'], {
                  "item" => host[0],
                  "host" => host[1]['host'],
                  "status" => checked,
                  "timestamp" => now
                })
              end
            end
          end
          host[1]['status'] = checked
          host[1]['lastcheck'] = now
        end
      end
      self.gendata('monitor', @@HOSTDATA)
      sleep(@@TIMEOUT)
    end
  end

  ## Remove activity lock-file
  def self.rmstatus!
    if File.exist?(@@STATUSFILE)
      FileUtils.rm(@@STATUSFILE)
    end
  end

  # public methods section
  public

  ## Start monitoring service
  def self.start!
    # remove lock-file if the previous session hasn't been ended correctly
    self.rmstatus!
    if self.config_load? and self.hosts_load?
      FileUtils.touch(@@STATUSFILE)
      self.log("Monitoring process started")
      self.watch!
    else
      self.log("Failed to start monitoring")
    end
  end

  ## Stop monitoring service
  def self.stop!
    self.rmstatus!
    self.log("Monitoring process stopped")
  end

end
